### PR TITLE
Replaced Dummy-Test-Objects with Mocks

### PR DESCRIPTION
**Pull request by RoiEXLab**
This replaces Objects like DummyTripleADisplay with mocks

---
This change is [Reviewable](https://reviewable.io/reviews/triplea-game/triplea/906)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -57,6 +57,7 @@ dependencies {
 	compile 'commons-httpclient:commons-httpclient:3.1'
 	compile 'org.apache.commons:commons-math3:3.6.1'
 	compile 'org.mindrot:jbcrypt:0.3m'
+	compile 'org.mockito:mockito-all:2.0.2-beta'
```
Comment (RoiEXLab) [inline build.gradle]: Had to insert this, because for whatever reason non-test classes are reffering to test cases so mockito needs to be available at non-test compile time

**Comment by RoiEXLab** [comment]
Interesting... Eclipse's JUnit Tool has no problems mocking anonymous inner classes.
Gradle has, which is kinda good - much cleaner code now...

**Comment by ron-murhammer** [comment]
@RoiEXLab This seems like a bad idea to me. I don't think using Mockito mocks in actual code is a good idea. At least I've never seen it done before.

**Comment by RoiEXLab** [comment]
@ron-murhammer agree with that... I will try to remove the usage in "normal" code...(using nul contructors etc.)

File: src/games/strategy/net/ServerMessenger.java [hunk 2]
```diff
@@ -52,7 +52,7 @@
   private final CopyOnWriteArrayList<IMessageListener> listeners = new CopyOnWriteArrayList<>();
   private final CopyOnWriteArrayList<IMessengerErrorListener> errorListeners =
       new CopyOnWriteArrayList<>();
-  private final CopyOnWriteArrayList<IConnectionChangeListener> connectionListeners =
+  protected final CopyOnWriteArrayList<IConnectionChangeListener> connectionListeners =
```
Comment (RoiEXLab) [inline src/games/strategy/net/ServerMessenger.java]: Quick reminder for me.. I need to revert this change since this is a result of testing wich didn't worked

File: src/games/strategy/triplea/TripleA.java [hunk 3]
```diff
@@ -120,8 +120,13 @@ public void startGame(final IGame game, final Set<IGamePlayer> players, final bo
       } else {
         headlessFrameUI = null;
       }
-      display = new DummyTripleADisplay(headlessFrameUI);
-      soundChannel = new DummySoundChannel();
+      display = new TripleADisplay(mock(TripleAFrame.class)){
```
Comment (RoiEXLab) [inline src/games/strategy/triplea/TripleA.java]: It's safe to call the constructor with null instead of a mock since I'm overriding the used method

File: src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java [hunk 4]
```diff
@@ -38,7 +39,7 @@ protected void launchInNewThread(final Component parent) {
     ServerGame game = null;
     try {
       m_gameData.doPreGameStartDataModifications(m_playerListing);
-      final IServerMessenger messenger = new DummyMessenger();
+      final IServerMessenger messenger = mock(IServerMessenger.class);
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java]: A `DummyMessenger` is a 'do-nothing' object. That is different from a mock.
The need for a `DummyMessenger` is actually a code smell. It indicates a lot of code that should not be called, but is. Secondly, it looks to be an inappropriate abstraction. It is overly broad and requires dependencies that are not always used.

Comment (DanVanAtta) [inline src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java]: Is this code covered by any tests? Is it only called from test code? Alternatively we need to do manual testing, which is expensive and quickly no longer a solution, or we add unit tests (or remove code if something is *only* called by test)

Comment (DanVanAtta) [inline src/games/strategy/engine/framework/startup/launcher/LocalLauncher.java]: To actually remove the DummyMessenger here, you'll need to remove all dependencies on the `IServerMessenger messenger` object in this method.

File: src/games/strategy/triplea/TripleA.java (hunk 3 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/TripleA.java]: This is confusing. Do not use mocks in 'production' code, test code only.

File: build.gradle (hunk 1 above)
Comment (DanVanAtta) [inline build.gradle]: Please remove and do the necessary to fix any errors - mocks are only for test code support.

File: test/games/strategy/triplea/delegate/MockObjects.java [hunk 5]
```diff
@@ -83,19 +83,37 @@ public Unit answer(InvocationOnMock invocation) throws Throwable {
...
   public static IServerMessenger getDummyMessenger() {
-    IServerMessenger messenger = null;
-    try {
-      messenger = spy(new ServerMessenger("", 0){
-        @Override
-        public void removeConnection(INode node){
-          for (final IConnectionChangeListener listener : connectionListeners) {
-            listener.connectionRemoved(node);
-          }
+    IServerMessenger messenger = mock(IServerMessenger.class);
```
Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/MockObjects.java]: Points for creativity. The way I see to get rid of the dummy messenger is to remove calls to `public static IServerMessenger getDummyMessenger() {` altogether, and then remove the method. The upstream code would need to deal with cases where there is no server messenger object. Likely the abstractions are not quite right, and too much coupling means we wind up creating a dummy version of this object, even though we could skip unnecessary computation.

File: test/games/strategy/engine/data/TestDelegateBridge.java [hunk 6]
```diff
@@ -30,7 +31,7 @@
   private PlayerID m_id;
   private String m_stepName = "no name specified";
   private IDisplay m_dummyDisplay;
-  private final DummySoundChannel m_soundChannel = new DummySoundChannel();
+  private final ISound m_soundChannel = mock(ISound.class);
```
Comment (DanVanAtta) [inline test/games/strategy/engine/data/TestDelegateBridge.java]: `TestDelegateBridge` should also be a mock. You may wind up definding something that looks like this:
```
TestDelegateBridge mockBridge
ISound mockSound;
:
:
expect(mockBridge.getSound().andReturn(mocksound);
expect(mockSound.doSomething()).andReturn(something);
```

File: test/games/strategy/engine/message/RemoteMessengerTest.java [hunk 7]
```diff
@@ -30,7 +31,7 @@
   @Before
   public void setUp() throws Exception {
     // simple set up for non networked testing
-    m_messenger = new DummyMessenger();
+    m_messenger = MockObjects.getDummyMessenger();
```
Comment (DanVanAtta) [inline test/games/strategy/engine/message/RemoteMessengerTest.java]: Should not need a utility for creating mock objects.

```
IMessenger mockMessenger
```

Then create expectations in the @Test methods. Copy/pasted expectations is a bad thing, those might go into a util class if fixing the source code directly is really difficult.

Comment (DanVanAtta) [inline test/games/strategy/engine/message/RemoteMessengerTest.java]: There is a lot of value in having self-contained tests. Avoiding external test  data or setup is very useful when debugging test failures.

File: test/games/strategy/triplea/delegate/BattleCalculatorTest.java [hunk 8]
```diff
@@ -139,18 +144,24 @@ public void testAACasualtiesLowLuckMixedWithChooseAACasualties() {
...
-        final List<Unit> selected = Match.getNMatches(selectFrom, count, Matches.UnitIsStrategicBomber);
-        return new CasualtyDetails(selected, new ArrayList<>(), false);
-      }
-    });
+    TripleAPlayer player = MockObjects.getDummyPlayer();
+    when(player.selectCasualties(any(Collection.class), any(Map.class), any(int.class), any(String.class),
+        any(DiceRoll.class), any(PlayerID.class), any(Collection.class), any(PlayerID.class), any(Collection.class),
+        any(boolean.class), any(Collection.class),
+        any(CasualtyList.class), any(GUID.class), any(Territory.class), any(boolean.class)))
+            .thenAnswer(new Answer<CasualtyDetails>() {
```
Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/BattleCalculatorTest.java]: `thenReturn`is preferred, `thenAnswer` is a bit of a last resort. 
In this case, the parameter value can likely be known, in which case we can expect an exact value, and return an exact value that we have constructed in the `@Setup`

File: test/games/strategy/triplea/delegate/BattleCalculatorTest.java [hunk 9]
```diff
@@ -139,18 +144,24 @@ public void testAACasualtiesLowLuckMixedWithChooseAACasualties() {
...
-          final PlayerID hit, final Collection<Unit> friendlyUnits, final PlayerID enemyPlayer,
-          final Collection<Unit> enemyUnits, final boolean amphibious, final Collection<Unit> amphibiousLandAttackers,
-          final CasualtyList defaultCasualties, final GUID battleID, final Territory battlesite,
-          final boolean allowMultipleHitsPerUnit) {
-        final List<Unit> selected = Match.getNMatches(selectFrom, count, Matches.UnitIsStrategicBomber);
-        return new CasualtyDetails(selected, new ArrayList<>(), false);
-      }
-    });
+    TripleAPlayer player = MockObjects.getDummyPlayer();
+    when(player.selectCasualties(any(Collection.class), any(Map.class), any(int.class), any(String.class),
```
Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/BattleCalculatorTest.java]: Avoid expecting `any` object. Try to match exact values or instances. For some of these I can see it being difficult to do otherwise.

File: test/games/strategy/triplea/delegate/BattleCalculatorTest.java [hunk 10]
```diff
@@ -178,18 +190,24 @@ public void testAACasualtiesLowLuckMixedWithChooseAACasualtiesRoll() {
...
+        any(DiceRoll.class), any(PlayerID.class), any(Collection.class), any(PlayerID.class), any(Collection.class),
+        any(boolean.class), any(Collection.class),
+        any(CasualtyList.class), any(GUID.class), any(Territory.class), any(boolean.class)))
+            .thenAnswer(new Answer<CasualtyDetails>() {
```
Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/BattleCalculatorTest.java]: DRY violation, but you should likely be using `thenReturn` anyways and returning test constants.

File: test/games/strategy/triplea/delegate/MockObjects.java [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+              return potentialTargets.iterator().next();
+            }
+            return typicalFactories.iterator().next();
+          }
+
+        });
+    return dummyPlayer;
+  }
+
+  public static IServerMessenger getDummyMessenger() {
```
Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/MockObjects.java]: Needing the same mock object in lot of tests indicates either lots of coupling, and/or we have redundant tests that keep constructing the same objects. 

Now that the `getDummyMessenger` is in test code, IServerMessenger should be mocked entirely. That is likely to be a big change, more than you want to bite off. 

Thoughts?

**Comment by DanVanAtta** [comment]
Got about half way, some significant things need to be updated, no mocks in test code. This is likely going to be biting off more than a person would want to chew. Some of the `dummy` class removals look pretty complete. IMO break them up and polish them so we can lock in that work. Then tackle the dummy classes one at a time.

File: test/games/strategy/engine/data/TestDelegateBridge.java (hunk 6 above)
Comment (RoiEXLab) [inline test/games/strategy/engine/data/TestDelegateBridge.java]: I know, but as experienced before, the Test delegate is used very often. And is used deep in the test logic... Will do that in some of the next PRs

**Comment by RoiEXLab** [comment]
@DanVanAtta Did a "partial revert" on this e.g. I readded a "Dummy class" called Headless instead...
I noticed, that dummy classes are to 90% only referenced from other dummy classes, so I think this code is going to be removed, once the remaining dummy delegatebridgecode is being removed...

File: build.gradle [hunk 12]
```diff
@@ -64,8 +64,8 @@ dependencies {
 
 	testCompile 'junit:junit:4.12'
 	testCompile 'org.apache.derby:derby:10.12.1.1'
-	testCompile 'org.hamcrest:hamcrest-all:1.3'
-	testCompile 'org.mockito:mockito-all:2.0.2-beta'
+	testCompile 'org.hamcrest:java-hamcrest:2.0.0.0'
+	testCompile 'org.mockito:mockito-core:2.0.82-beta'
```
Comment (RoiEXLab) [inline build.gradle]: Updated mockito in Order to fix something which didn't work... Kept for the any() thing...
Hamcrest update is just btw

**Comment by RoiEXLab** [comment]
@DanVanAtta @ron-murhammer updated!

**Comment by DanVanAtta** [comment]
LGTM
